File: project/GO_SDK/samples/C#/ReceiveProfile/ReceiveProfile.cs
/*
 * ReceiveProfile.cs
 * 
 * Gocator 2000/2300 C# Sample
 * Copyright (C) 2013-2021 by LMI Technologies Inc.
 * 
 * Licensed under The MIT License.
 * Redistributions of files must retain the above copyright notice.
 *
 * Purpose: Connect to Gocator system and receive range data in Profile Mode and translate to engineering units (mm). Gocator must be in Profile Mmode.
 * Ethernet output for the profile data must be enabled.
 */

using System;
using System.Runtime.InteropServices;
using Lmi3d.GoSdk;
using Lmi3d.Zen;
using Lmi3d.Zen.Io;
using Lmi3d.GoSdk.Messages;

static class Constants
{
    public const string SENSOR_IP = "192.168.1.10"; // IP of the sensor used for sensor connection GoSystem_FindSensorByIpAddress() call.
}

namespace ReceiveProfile
{
    public class DataContext
    {
        public Double xResolution;
        public Double zResolution;
        public Double xOffset;
        public Double zOffset;
        public uint serialNumber;
    }

    public struct ProfilePoint
    {
        public double x;
        public double z;
        byte intensity;
    }

    public struct GoPoints
    {
        public Int16 x;
        public Int16 y;
    }

    class ReceiveProfile
    {
        static int Main(string[] args)
        {
            try
            {
                KApiLib.Construct();
                GoSdkLib.Construct();
                GoSystem system = new GoSystem();
                GoSensor sensor;
                KIpAddress ipAddress = KIpAddress.Parse(Constants.SENSOR_IP);
                GoDataSet dataSet = new GoDataSet();
                sensor = system.FindSensorByIpAddress(ipAddress);
                sensor.Connect();
                system.EnableData(true);
                system.Start();
                dataSet = system.ReceiveData(30000000);
                DataContext context = new DataContext();
                for (UInt32 i = 0; i < dataSet.Count; i++)
                {
                    GoDataMsg dataObj = (GoDataMsg)dataSet.Get(i);
                    switch (dataObj.MessageType)
                    {
                        case GoDataMessageType.Stamp:
                            {
                                GoStampMsg stampMsg = (GoStampMsg)dataObj;
                                for (UInt32 j = 0; j < stampMsg.Count; j++)
                                {
                                    GoStamp stamp = stampMsg.Get(j);
                                    Console.WriteLine("Frame Index = {0}", stamp.FrameIndex);
                                    Console.WriteLine("Time Stamp = {0}", stamp.Timestamp);
                                    Console.WriteLine("Encoder Value = {0}", stamp.Encoder);
                                }
                            }
                            break;

                        case GoDataMessageType.UniformProfile:
                            {
                                GoUniformProfileMsg profileMsg = (GoUniformProfileMsg)dataObj;
                                Console.WriteLine("  Resampled Profile Message batch count: {0}", profileMsg.Count);
                                for (UInt32 k = 0; k < profileMsg.Count; ++k)
                                {
                                    int validPointCount = 0;
                                    int profilePointCount = profileMsg.Width;
                                    Console.WriteLine("    Item[{0}]: Profile data ({1} points)", k, profileMsg.Width);
                                    context.xResolution = (double)profileMsg.XResolution / 1000000;
                                    context.zResolution = (double)profileMsg.ZResolution / 1000000;
                                    context.xOffset = (double)profileMsg.XOffset / 1000;
                                    context.zOffset = (double)profileMsg.ZOffset / 1000;

                                    short[] points = new short[profilePointCount];
                                    ProfilePoint[] profileBuffer = new ProfilePoint[profilePointCount];
                                    IntPtr pointsPtr = profileMsg.Data;
                                    Marshal.Copy(pointsPtr, points, 0, points.Length);

                                    for (UInt32 arrayIndex = 0; arrayIndex < profilePointCount; ++arrayIndex)
                                    {
                                        if (points[arrayIndex] != -32768)
                                        {
                                            profileBuffer[arrayIndex].x = context.xOffset + context.xResolution * arrayIndex;
                                            profileBuffer[arrayIndex].z = context.zOffset + context.zResolution * points[arrayIndex];
                                            validPointCount++;
                                        }
                                        else
                                        {
                                            profileBuffer[arrayIndex].x = context.xOffset + context.xResolution * arrayIndex;
                                            profileBuffer[arrayIndex].z = -32768;
                                        }
                                    }
                                    Console.WriteLine("Received {0} Range Points", profilePointCount);
                                    Console.WriteLine("Valid Points {0}", validPointCount);
                                }
                            }
                            break;

                        case GoDataMessageType.ProfilePointCloud:
                            {
                                GoProfilePointCloudMsg profileMsg = (GoProfilePointCloudMsg)dataObj;
                                Console.WriteLine("  Profile Message batch count: {0}", profileMsg.Count);
                                for (UInt32 k = 0; k < profileMsg.Count; ++k)
                                {
                                    int validPointCount = 0;
                                    long profilePointCount = profileMsg.Width;
                                    Console.WriteLine("    Item[{0}]: Profile data ({1} points)", i, profileMsg.Width);
                                    context.xResolution = profileMsg.XResolution / 1000000;
                                    context.zResolution = profileMsg.ZResolution / 1000000;
                                    context.xOffset = profileMsg.XOffset / 1000;
                                    context.zOffset = profileMsg.ZOffset / 1000;
                                    GoPoints[] points = new GoPoints[profilePointCount];
                                    ProfilePoint[] profileBuffer = new ProfilePoint[profilePointCount];
                                    int structSize = Marshal.SizeOf(typeof(GoPoints));
                                    IntPtr pointsPtr = profileMsg.Data;
                                    for (UInt32 array = 0; array < profilePointCount; ++array)
                                    {
                                        IntPtr incPtr = new IntPtr(pointsPtr.ToInt64() + array * structSize);
                                        points[array] = (GoPoints)Marshal.PtrToStructure(incPtr, typeof(GoPoints));
                                    }

                                    for (UInt32 arrayIndex = 0; arrayIndex < profilePointCount; ++arrayIndex)
                                    {
                                        if (points[arrayIndex].x != -32768)
                                        {
                                            profileBuffer[arrayIndex].x = context.xOffset + context.xResolution * points[arrayIndex].x;
                                            profileBuffer[arrayIndex].z = context.zOffset + context.zResolution * points[arrayIndex].y;
                                            validPointCount++;
                                        }
                                        else
                                        {
                                            profileBuffer[arrayIndex].x = -32768;
                                            profileBuffer[arrayIndex].z = -32768;
                                        }
                                    }
                                    Console.WriteLine("Received {0} Range Points", profilePointCount);
                                    Console.WriteLine("Valid Points {0}", validPointCount);
                                }
                            }
                            break;

                        case GoDataMessageType.ProfileIntensity:
                            {
                                GoProfileIntensityMsg profileMsg = (GoProfileIntensityMsg)dataObj;
                                Console.WriteLine("  Profile Intensity Message batch count: {0}", profileMsg.Count);
                                for (UInt32 k = 0; k < profileMsg.Count; ++k)
                                {
                                    byte[] intensity = new byte[profileMsg.Width];
                                    IntPtr intensityPtr = profileMsg.Data;
                                    Marshal.Copy(intensityPtr, intensity, 0, intensity.Length);
                                }
                            }
                            break;
                    }
                }
                system.Stop();
            }
            catch (KException ex)
            {
                Console.WriteLine("Error: {0}", ex.ToString());
            }
            // wait for ENTER key
            Console.WriteLine("\nPress ENTER to continue");
            while (Console.ReadKey().Key != ConsoleKey.Enter) { }

            return 1;
        }
    }
}

File: project/GO_SDK/samples/VB.NET/ReceiveProfile/ReceiveProfile.vb
'ReceiveProfile.vb

'Gocator 2300 VB.NET Sample Code
'Copyright (C) 2013, 2018 by LMI Technologies Inc.

'Licensed under The MIT License.
'Redistributions of files must retain the above copyright notice.

'Purpose: Connect to Gocator sensor and receive profile data under Profile Mode.
'         Make sure profile Top output is enabled

Imports Lmi3d.GoSdk
Imports Lmi3d.GoSdk.Messages
Imports Lmi3d.Zen
Imports Lmi3d.Zen.Io

Module ReceiveProfile
    Public Const IP As String = "192.168.1.10"
    Public Const TIMEOUT As UInteger = 100000
    Public Const OUT_OF_RANGE = -32768

    Function Main() As Integer
        Try
            Dim dataObj As GoDataMsg
            Dim gocator As GoSensor = Nothing

            KApiLib.Construct()
            GoSdkLib.Construct()

            Dim gsystem As New GoSystem

            Dim ipAddress As KIpAddress = KIpAddress.Parse(IP)
            gocator = gsystem.FindSensorByIpAddress(ipAddress)

            gocator.Connect()
            Console.WriteLine("GoSdk   Release : {0}", gsystem.SdkVersion.Format())
            Console.WriteLine("Gocator Firmware: {0}", gocator.FirmwareVersion.Format())

            gocator.EnableData(True)

            gocator.Stop()

            gocator.Start()

            ' read data from sensor
            Dim dataset As GoDataSet = gsystem.ReceiveData(TIMEOUT)
            Dim itemCount = dataset.Count()
            If (itemCount > 0) Then
                ' each result can have multiple data items
                ' loop through all items in result message
                For i As Integer = 0 To itemCount - 1
                    dataObj = dataset.Get(i)

                    ' retrieve GoStamp message
                    Select Case dataObj.MessageType
                        Case GoDataMessageType.Stamp
                            Dim stampMsg As GoStampMsg = dataObj
                            For j As Integer = 0 To stampMsg.Count - 1
                                Dim stamp As GoStamp = stampMsg.Get(j)
                                Console.WriteLine("Frame Index = {0}", stamp.FrameIndex)
                                Console.WriteLine("Time Stamp = {0}", stamp.Timestamp)
                                Console.WriteLine("Encoder Value = {0}", stamp.Encoder)
                            Next

                        Case GoDataMessageType.Profile Or GoDataMessageType.ProfilePointCloud
                            Dim profileMsg As GoProfilePointCloudMsg = dataObj
                            Dim profilePoints(profileMsg.Count, profileMsg.Width) As Data.KPoint16s
                            For j As Integer = 0 To profileMsg.Count - 1
                                For k As Integer = 0 To profileMsg.Width - 1
                                    profilePoints(j, k) = profileMsg.Get(k)
                                    If profilePoints(j, k).Y <> OUT_OF_RANGE Then
                                        ' show only valid values
                                        Console.WriteLine("Profile[{0}][{1}] X {2} Z {3}", j, k, profilePoints(j, k).X, profilePoints(j, k).Y)
                                    End If
                                Next
                            Next

                        Case GoDataMessageType.UniformProfile Or GoDataMessageType.ResampledProfile
                            Dim profileResampledMsg As GoUniformProfileMsg = dataObj
                            Dim profilesCount As Long = profileResampledMsg.Count
                            Dim profileWidth As Short = profileResampledMsg.Width
                            Dim profilePoints(profilesCount, profileWidth) As Int16
                            For j As Integer = 0 To profilesCount - 1
                                For k As Integer = 0 To profileWidth - 1
                                    profilePoints(j, k) = profileResampledMsg.Get(k)
                                    If profilePoints(j, k) <> OUT_OF_RANGE Then
                                        Console.WriteLine("Profile[{0}][{1}]  {2} ", j, k, profilePoints(j, k))
                                    End If
                                Next
                            Next

                        Case GoDataMessageType.ProfileIntensity
                            Dim profileIntensityMsg As GoProfileIntensityMsg
                            profileIntensityMsg = dataObj
                            Dim profilesCount As Long = profileIntensityMsg.Count
                            Dim profileWidth As Integer = profileIntensityMsg.Width
                            Dim profileIntensities(profilesCount, profileWidth) As Byte
                            For j As Long = 0 To profilesCount - 1
                                For k As Long = 0 To profileWidth - 1
                                    profileIntensities(j, k) = profileIntensityMsg.Get(j, k)
                                    If profileIntensities(j, k) > 0 Then
                                        Console.WriteLine("Profile[{0}][{1}] Intensity {2}", j, k, profileIntensities(j, k))
                                    End If
                                Next
                            Next

                    End Select

                Next
            End If
            dataset.Destroy()
            gocator.Stop()

            Console.WriteLine(Environment.NewLine)
            'wait for ESC key
            Console.WriteLine("Press ESCAPE to Exit")
            Do While Not Console.ReadKey(True).Key = ConsoleKey.Escape
            Loop
            gocator.Destroy()
        Catch ex As Exception
            Console.WriteLine("Error: {0}.....Press ESCAPE to Exit", ex.ToString())
            Do While Not Console.ReadKey(True).Key = ConsoleKey.Escape
        Loop
        Return KStatus.Error
        Exit Function
        End Try
        Return KStatus.Ok
    End Function

End Module